fdbserver/include/fdbserver/DataDistribution.actor.h

```diff
@@ -340,6 +342,154 @@ struct ShardTrackedData {
+
+// Determines the maximum shard size based on the size of the database
+int64_t getMaxShardSize(double dbSizeEstimate);
+
+#ifndef __INTEL_COMPILER
+#pragma endregion
+#endif
+
+/////////////////////////////// DD Components //////////////////////////////////////
+#ifndef __INTEL_COMPILER
+#pragma region DD Components
+#endif
+class DDTeamCollection;
+
+// send request/signal to TeamCollection through interface
+// call synchronous method from components outside DDTeamCollection
+struct ITeamCollection {
+	struct Interface {
+		PromiseStream<GetTeamRequest> getTeam;
+	};
+	virtual ~ITeamCollection() {}
+};
+
+// send request/signal to DDTracker through interface
+// call synchronous method from components outside DDTracker
+struct IDDTracker {
+	struct Interface {
+		Promise<Void> readyToStart;
+		PromiseStream<GetMetricsRequest> getShardMetrics;
+		PromiseStream<GetTopKMetricsRequest> getTopKMetrics;
+		PromiseStream<GetMetricsListRequest> getShardMetricsList;
+		PromiseStream<KeyRange> restartShardTracker;
+		PromiseStream<Promise<int64_t>> getAverageShardBytes; // FIXME(xwang): change it to a synchronous call
+	};
+	virtual double getAverageShardBytes() = 0;
+	virtual ~IDDTracker() = 0;
+};
+
+// send request/signal to DDQueue through interface
+// call synchronous method from components outside DDQueue
+struct IDDQueue {
+	struct Interface {
+		PromiseStream<RelocateShard> relocationProducer, relocationConsumer;
+		PromiseStream<Promise<int>> getUnhealthyRelocationCount; // FIXME(xwang): change it to a synchronous call
+	};
+	virtual int getUnhealthyRelocationCount() = 0;
+	virtual ~IDDQueue() = 0;
+};
+
+// The common info shared by all DD components. Normally the DD components should share the reference to the same
+// context.
+struct DDContext : public ReferenceCounted<DDContext> {
+	// FIXME(xwang) mark fields privates
+	// private:
+	std::shared_ptr<DDEnabledState>
+	    ddEnabledState; // Note: don't operate directly because it's shared with snapshot server
+	IDDTracker::Interface trackerInterface;
+	IDDQueue::Interface queueInterface;
+	// public:
+	UID ddId;
+	MoveKeysLock lock;
+	bool trackerCancelled = false;
+	DatabaseConfiguration configuration;
+
+	Reference<ShardsAffectedByTeamFailure> shardsAffectedByTeamFailure;
+	Reference<AsyncVar<bool>> processingUnhealthy, processingWiggle;
+
+	DDContext() = default;
+
+	DDContext(UID id, std::shared_ptr<DDEnabledState> ddEnabledState)
+	  : ddEnabledState(std::move(ddEnabledState)), ddId(id),
+	    shardsAffectedByTeamFailure(new ShardsAffectedByTeamFailure), processingUnhealthy(new AsyncVar<bool>(false)),
+	    processingWiggle(new AsyncVar<bool>(false)) {}
+
+	void initTeamCollectionStates();
+
+	void proposeRelocation(const RelocateShard& rs) const { return queueInterface.relocationProducer.send(rs); }
+
+	void restartShardTrackerAsync(KeyRange keys) const { return trackerInterface.restartShardTracker.send(keys); }
```

Comment:
These do not belong in this class, IMHO.

Reply:
I think one of the responsibilities of `DDContext` is being a mediator that simplifies the communication between components, and it knows all other interfaces so it's somehow natural to provide those methods rather than let each component explicitly retrieve interfaces and then use them. I'm not sure whether there's a better place for this method, but for now, since this class is not used yet so it still can be changed, I prefer to leave it like this.